### PR TITLE
installation-telemetry: add customer ID

### LINKED ISSUE #10183: Expand Self-Hosted Telemetry With Customer ID
## Is your feature request related to a problem? Please describe
Currently, all of the data we receive from each Self-Hosted Gitpod instance (assuming the user did not opt-out) is anonymised. (For context, the only data we currently receive is at the aggregated level: Gitpod version, number of users, number of instances and number of workspaces). This means that although we can extract general insights from this data, we cannot specifically help individual customers as we have no way to identify them. 

## Describe the behaviour you'd like
- Add the capability to send the license ID or some other form of unique identifier of the owner of a given self-hosted instance with each data package
- Allow users to specifically opt-out of this identifiable info (@gtsiolis I think it would be great to get your design take on such a two-layer opt out)
- Make sure this is communicated in honesty to users that we are doing this and why we are doing this. Explain to them how they can opt-out. 

## Additional context
Internal slack threads ([1](https://gitpod.slack.com/archives/C01LDA28WV6/p1653053314293259),[2](https://gitpod.slack.com/archives/C01KLC56NP7/p1653042186959809))

### PR DESCRIPTION
**Pull request by adrienthebo**
## Description

This pull request exposes the GitPod Customer ID field in installation telemetry.

## Related Issue(s)

This issue resolves #10183.

## How to test

The `customerID` field will be available in the Gitpod Admin/Settings page, displayed in the installation-telemetry job output, and will be present in the segment.io telemetry.

<details>
<summary>Telemetry Disabled</summary>

**Dashboard: Admin/Settings:**

Dashboard: Admin/Settings, telemetry disabled

**Installation telemetry job:**

```json
 {
  "level": "info",
  "message": "installation-telemetry is not permitted to send - exiting",
  "serviceContext": {
    "service": "installation-telemetry",
    "version": "commit-2b1c7e10da04762651d753db7b1fb1b9954b3e70"
  },
  "severity": "INFO",
  "time": "2022-06-24T15:28:09Z"
}
```

</details>

<details>
<summary>Telemetry enabled, Customer ID disabled</summary>

**Dashboard: Admin/Settings:**

Screen Shot 2022-06-24 at 8 35 23 AM

**Installation Telemetry job:**

```json
{
  "level": "info",
  "message": "installation-telemetry has successfully sent data - exiting",
  "serviceContext": {
    "service": "installation-telemetry",
    "version": "commit-2b1c7e10da04762651d753db7b1fb1b9954b3e70"
  },
  "severity": "INFO",
  "telemetry": {
    "userId": "cfcc2cfa-9ac6-482f-879e-f327c260c533",
    "event": "Installation telemetry",
    "timestamp": "0001-01-01T00:00:00Z",
    "properties": {
      "totalInstances": 0,
      "totalUsers": 1,
      "totalWorkspaces": 0,
      "version": "gitpod-custom-alt-10183-sh-telem-license.30"
    }
  },
  "time": "2022-06-24T15:36:05Z"
}
```

</details>

<details>
<summary>Telemetry enabled, Customer ID enabled, empty customer ID</summary>

**Dashboard: Admin/Settings:**
Screen Shot 2022-06-24 at 8 38 03 AM

**Installation Telemetry job:**

```json
{
  "level": "info",
  "message": "installation-telemetry has successfully sent data - exiting",
  "serviceContext": {
    "service": "installation-telemetry",
    "version": "commit-2b1c7e10da04762651d753db7b1fb1b9954b3e70"
  },
  "severity": "INFO",
  "telemetry": {
    "userId": "cfcc2cfa-9ac6-482f-879e-f327c260c533",
    "event": "Installation telemetry",
    "timestamp": "0001-01-01T00:00:00Z",
    "properties": {
      "customerID": "",
      "totalInstances": 0,
      "totalUsers": 1,
      "totalWorkspaces": 0,
      "version": "gitpod-custom-alt-10183-sh-telem-license.30"
    }
  },
  "time": "2022-06-24T15:39:50Z"
}
```

</details>

<details>
<summary>Installation telemetry enabled, customer ID enabled, customer ID set</summary>

**Dashboard: Settings/Admin:**

Screen Shot 2022-06-24 at 8 47 52 AM

**Installation telemetry job:**

```json
{
  "level": "info",
  "message": "installation-telemetry has successfully sent data - exiting",
  "serviceContext": {
    "service": "installation-telemetry",
    "version": "commit-2b1c7e10da04762651d753db7b1fb1b9954b3e70"
  },
  "severity": "INFO",
  "telemetry": {
    "userId": "cfcc2cfa-9ac6-482f-879e-f327c260c533",
    "event": "Installation telemetry",
    "timestamp": "0001-01-01T00:00:00Z",
    "properties": {
      "customerID": "Tessier Ashpool S.A.",
      "totalInstances": 0,
      "totalUsers": 1,
      "totalWorkspaces": 0,
      "version": "gitpod-custom-alt-10183-sh-telem-license.30"
    }
  },
  "time": "2022-06-24T15:48:56Z"
}
```

</details>

## Release Notes

```release-note
[self-hosted] Installation telemetry optionally includes the Gitpod customer ID
```

## Documentation

This change is user facing, and when enabled adds information that removes the pseudononymous guarantees in our documentation. We'll amend that in a separate pull request against the website.

### CODE REVIEW

**Comment by adrienthebo** [comment]
This looks like the telemetry data provider needs to be relocated into `components/server/ee/src/installation-admin`, but the overall approach is ready for review.

**Review by nandajavarma — APPROVED** [review]
Works great! Great work Adrien! :star_struck:

**Comment by lucasvaltl** [comment]
Hey @adrienthebo, we originally discussed two implementation paths here - one with the license ID, and one with a newly added customer ID as a (hidden) field in the replicated license. We had plan was to use the customer ID (reasoning was outlined in [this comment](https://github.com/gitpod-io/gitpod/issues/10183#issuecomment-1135491136)) but I had failed to transfer this state into the main issue description (Sorry!). Is it possible to switch the implementation to use customer id instead of the license ID?

File: components/licensor/ee/cmd/sign.go [hunk 1]
```diff
@@ -90,6 +92,7 @@ func init() {
 	rootCmd.AddCommand(signCmd)
 
 	signCmd.Flags().String("domain", "", "domain for which the license is valid")
+	signCmd.Flags().String("customer-id", "", "analytics customer ID")
```
Comment (adrienthebo) [inline components/licensor/ee/cmd/sign.go:95]: I'm on the fence about whether it's worthwhile to expose customer-id in the licensor `sign` command. Leaving this value out seems a bit surprising but we're not planning on using this heavily going forward. I'd appreciate an outside opinion - should we keep this internal only or plumb it through to the command?

File: components/licensor/ee/pkg/licensor/licensor.go [hunk 2]
```diff
@@ -53,6 +53,9 @@ type LicensePayload struct {
 
 	// Seats == 0 means there's no seat limit
 	Seats int `json:"seats"`
+
+	// CustomerID is used to identify installations in installation analytics
+	CustomerID string `json:"customerID"`
```
Comment (adrienthebo) [inline components/licensor/ee/pkg/licensor/licensor.go:58]: There are two hard problems in computer science; naming, cache invalidation, and off by one errors.

Within Replicated we have a custom `levelId` field; at present I'm naming this field `customerId` to match that pattern. However within this implementation we use `customerID` with an upper cased D. If anyone has strong opinions I'm happy to adjust this.

File: components/licensor/ee/pkg/licensor/replicated.go [hunk 3]
```diff
@@ -118,6 +118,9 @@ func newReplicatedEvaluator(client *http.Client) (res *Evaluator) {
 
 		case "seats":
 			lic.Seats = int(i.Value.(float64))
+
+		case "customerId":
```
Comment (adrienthebo) [inline components/licensor/ee/pkg/licensor/replicated.go:122]: There are two hard problems in computer science; naming, cache invalidation, and off by one errors.

Within Replicated we have a custom `levelId` field; at present I'm naming this field `customerId` to match that pattern. However within this implementation we use `customerID` with an upper cased D. If anyone has strong opinions I'm happy to adjust this.

**Comment by lucasvaltl** [comment]
This looks good - two things that are missing as per the original ticket:
- Allow users to specifically opt-out of this identifiable info (@gtsiolis I think it would be great to get your design take on such a two-layer opt out)
--> How much additional work do you think this is? The reason for including this was to avoid customers opting out completely just because we are now sending data that allows us to know who the owner of an installation is. I was thinking of having an additional checkbox that is ticked by default stating "Include Customer ID".  Happy to chat in sync about this - just reach out :) 

- Make sure this is communicated in honesty to users that we are doing this and why we are doing this. Explain to them how they can opt-out.
--> What are your plans here? We should at the very least update our docs which state that ["We do not collect any data that allows conclusions to be drawn about the owner of the installation."](https://www.gitpod.io/docs/self-hosted/latest/telemetry) and notify #t_customer_experience on slack :)

**Comment by corneliusludmann** [comment]
Just my 2 cents: Given that the customer ID is included only when we explicitly add it to the license (and we could ask the customers for consent before we do this), in my option we could add this in a separate PR (especially when we want to give George some time to give his input). What do you think?

**Comment by geropl** [comment]
@adrienthebo I was looking to review, but couldn't because the build seems to fail. :confused:

**Review by geropl — APPROVED** [review]
LGTM, tested and works as expected :+1:

**Pull request merged**